### DISCUSSION TITLE
检测一下各位绳匠的编程能力

### DISCUSSION THREAD

**Discussion by qwq1949** [share121/inter-knot#809] [其他]
## 题目背景
你有两个很智慧的朋友 **A** 和 **B**
他们发生了以下对话...
有两个不相等的正整数 _a_，_b_。
**A**  知道 _a + b_ ，**B** 知道 _a * b_ 。
**A** ：我不知道 _a_，_b_ 是什么，但我肯定你也不知道。
**B** ：你说这句话之前我不知道不代表我现在不知道。
你决定展示出你的实力，狠狠地展现实力！
你想知道 _a_，_b_ 是什么
可惜脑子不够了呢... 怎么办呢？
于是你掏出了超级大宝贝...~当当当当~
## 题目描述
根据对话给出 𝑎，𝑏 的值。
𝑎 和 𝑏 有很多对哦~ 
每一对一行，中间空格隔开。

### 输入格式
没有捏
### 输出格式
3 8（举个例子）

### 提示
数据范围 ： _2 ≤_ _a_，_b_ _≤ 114_

##### 样例解释：
A 知道 11，B 知道 24
满足 A 保证 B 不知道
也满足 B 知道 A 保证 B 不知道后知道
...反正满足题意就是了！
###### ps : 题目至少没查到一样的 如果有那一定是巧合！
-------
## 经传奇绳匠 Dave12138 大佬指正，题目已改正~
-------
###### 然后就多出来一问...
结果 **A** 又说 : 我现在也知道了捏~
那么 _a_，_b_ 又可能是什么呢？

**Comment by Lanxey** [comment]
转铅笔可得答案是C

**Comment by aiTunes-dev** [comment]
用的C++，帮我看看有什么问题？
#include <iostream>
#include <vector>
#include <utility> // for std::pair

// 函数用于检查B不知道 a, b
bool checkBDoesNotKnow(int sum, int product) {
    std::vector<std::pair<int, int>> pairs;
    
    // 找到所有可能的 (a, b) 配对
    for (int i = 1; i <= 93; ++i) {
        for (int j = i + 1; j <= 93; ++j) {
            if (i + j == sum && i * j == product) {
                pairs.emplace_back(i, j);
            }
        }
    }

    // 如果找到的配对数量大于1，说明B不能确定具体是哪一对
    return pairs.size() > 1;
}

// 检查能否满足A的声明，并且B可以推导出答案
void findPairs(int sum, int product) {
    std::vector<std::pair<int, int>> validPairs;

    // 找到所有满足条件的配对
    for (int a = 1; a <= 93; ++a) {
        for (int b = a + 1; b <= 93; ++b) { // 确保 a，不等于 b
            if (a + b == sum && a * b == product) {
                validPairs.emplace_back(a, b);
            }
        }
    }

    // 验证有效配对列举
    for (const auto& pair : validPairs) {
        if (checkBDoesNotKnow(sum, product)) {
            std::cout << "Valid pair found: a = " << pair.first << ", b = " << pair.second << "\n";
        }
    }
}

int main() {
    int sum = 11;
    int product = 24;

    findPairs(sum, product);

    return 0;
}

**Reply by qwq1949** [reply]
为啥没输出捏qwq

**Comment by Dave-12138** [comment]
>用的C++，帮我看看有什么问题？

markdown里用一对```（键盘左上角，esc下面那个键）包裹起来（前三个后三个）就可以保持代码格式了，看起来就像这样：
```
document.addEventListener('click', e => {
    const elements = [].filter.call(document.body.querySelectorAll('*'), l => ![].includes.call(l.querySelectorAll('*'), e.target));
    elements.forEach(el => el.dispatchEvent(new Event('clickoutside')));
})
```

**Comment by khiqwq** [comment]
【灌水】
这个委托我替AI接了
这个问题是经典的数论与逻辑推理问题，考察了你对信息传递的理解。我们可以写一个程序来解决这个问题，通过穷举所有可能的 $a$ 和 $b$，然后根据对话的逻辑条件筛选出符合条件的解。

下面是C++代码的一个实现示例：

#### Cpp 代码示例 ####

```
#include <iostream>
#include <vector>
#include <set>

using namespace std;

// 检查给定的和 s 和乘积 p 是否满足条件
bool check(int a, int b, int s, int p) {
    for (int x = 2; x < s; ++x) {
        int y = s - x;
        if (y > 1 && x * y == p) {
            if (x != a && y != b && y != a && x != b) {
                return false;
            }
        }
    }
    return true;
}

int main() {
    vector<pair<int, int>> result;

    // 穷举所有可能的 a 和 b
    for (int a = 2; a <= 93; ++a) {
        for (int b = a + 1; b <= 93; ++b) {
            int s = a + b;
            int p = a * b;
            bool a_knows = true, b_knows = true;

            // 检查 A 是否知道 a 和 b 是什么
            for (int x = 2; x < s; ++x) {
                int y = s - x;
                if (y > 1 && x * y != p) {
                    a_knows = false;
                    break;
                }
            }

            // 如果 A 不知道
            if (!a_knows) {
                b_knows = check(a, b, s, p);
            }

            // 如果 B 在 A 说之前不知道，之后知道了
            if (!a_knows && b_knows) {
                result.push_back({a, b});
            }
        }
    }

    // 输出结果
    for (const auto& pair : result) {
        cout << pair.first << " " << pair.second << endl;
    }

    return 0;
}

```
#### Cpp 代码结束 ####

这个程序会输出所有满足条件的 $(a, b)$ 对。核心思想是在穷举可能的 $a$ 和 $b$ 后，通过检查 A 和 B 是否能够确定 $a$ 和 $b$ 的值来筛选出符合条件的组合。

你可以直接运行这个程序来获得所有的符合条件的解。

**Reply by qwq1949** [reply]
AI小笨蛋 答错喽🫠

**Comment by aaa00073** [comment]
答案是买两个西瓜

**Reply by qwq1949** [reply]
好聪明诶🫠

**Comment by BSNHD** [comment]
死去的信奥还在追杀我[悲]

**Comment by c1iied** [comment]
不懂

**Comment by Buer-Nahida** [comment]
# 写完了

```rust
fn main() {
    for a in 2..=93 {
        for b in 2..=93 {
            if a != b {
                continue;
            }
            let sum = a + b;
            let product = a * b;
            let possible_pair = possible_pairs(|a, b| a != b && a + b == sum);
            if possible_pair.len() > 1
                && possible_pairs(|a, b| a != b && a * b == product).len() > 1
            {
                let mut b_possible = false;
                for &(x, y) in &possible_pair {
                    if x * y == product {
                        b_possible = true;
                        break;
                    }
                }
                if b_possible {
                    println!("{} {}", a, b);
                }
            }
        }
    }
}

fn possible_pairs<T>(f: T) -> Vec<(i32, i32)>
where
    T: Fn(i32, i32) -> bool,
{
    let mut pairs = Vec::new();
    for a in 2..=93 {
        for b in 2..=93 {
            if f(a, b) {
                pairs.push((a, b));
            }
        }
    }
    pairs
}
```

**Reply by qwq1949** [reply]
第四行是 a == b 吧
不过好像还是不对qwq

**Reply by Buer-Nahida** [reply]
确实，最后修缮代码时整出来了好多bug
```rust
fn main() {
    for a in 2..=93 {
        for b in 2..=93 {
            if a == b {
                continue;
            }
            let sum = a + b;
            let product = a * b;
            let possible_pair = possible_pairs(|a, b| a != b && a + b == sum);
            if possible_pair.len() > 1
                && possible_pairs(|a, b| a != b && a * b == product).len() > 1
            {
                let mut b_possible = false;
                for &(x, y) in &possible_pair {
                    if x * y == product {
                        b_possible = true;
                        break;
                    }
                }
                if b_possible {
                    println!("{} {}", a, b);
                }
            }
        }
    }
}

fn possible_pairs<T>(f: T) -> Vec<(i32, i32)>
where
    T: Fn(i32, i32) -> bool,
{
    let mut pairs = Vec::new();
    for a in 2..=93 {
        for b in 2..=93 {
            if f(a, b) {
                pairs.push((a, b));
            }
        }
    }
    pairs
}

**Reply by qwq1949** [reply]
e..好像还是...不太对啊qwq

**Reply by Buer-Nahida** [reply]
ummm...难道是我理解错题目意思了？

**Comment by qwq1949** [comment]
###### 再给几对成立的 a 和 b 吧~ qwq
###### 2 49
###### 3 8
###### 3 32
这是单机绳网吗qwq 怎么都是AI （bushi） qwq

**Comment by YeMengXue** [comment]
中职的网页课还在追我，啊啊啊啊啊啊———

**Comment by uidsd604** [comment]
来个大佬写个python的，c++看不懂

**Comment by Qiling-XwX** [comment]
![FA33A880B9F6DFF5CD4433778AB8375A](https://github.com/user-attachments/assets/ba6a45f6-a891-4c1b-bd4c-22b8daa21dbd)

**Comment by Dave-12138** [comment]
没看懂题，是我漏了什么吗
我的理解：
A拿到11后，推测B可能拿到18、24、28、30
其中18=2\*9=3\*6
24=2\*12=3\*8=4\*6
28=2\*14=4\*7
30=2\*15=3\*10=5\*6
### 没有一个是有唯一解的，所以A说哈↗哈↘，你也不知道吧
B刚刚还在想自己拿到了24，所以A可能拿到了10、11、14
就听见A发出的b动静
B就想，如果A拿到10，则A眼里自己拿到的可能是16、21、24、25，而25一眼5\*5，他肯定不会发出b动静
如果A拿到14，则他眼里自己可能拿到24、33、40、45、48、49，一眼33=3\*11，肯定也不叫
### 就是11了，3\*8=24；3+8=11，稳了
于是B也叫了起来
A听见B的b动静，想到：
如果对方拿到18，对方猜自己应该是11、9
11怪叫，9有2\*7不怪叫
如果对方拿到24，对方猜自己10、11、14
10不叫，11叫，14不叫
如果对方拿到28，对方猜自己11、16
11叫，16有11\*5不叫
如果对方拿到30，对方猜自己11、13、17
11叫，13有2\*11不叫，17叫（30、42、52、60、66、70、72）

因此，如果B拿到18、24、28，他肯定能从b动静猜到A是11，30则不然
于是A知道了B拿的不是30，而是18、24、28其中之一
### 然后A就没信息继续排除了，当场开启死鸭子嘴硬形态

**Reply by qwq1949** [reply]
**难道这是法厄同？！！**

emmm...
 A 在看到自己的数之后能保证 B 不知道
所以在 A 角度需要保证拆开成 B 的数一定不会让 B 叫
符合的只有都不叫的 24 吧
###### （ 我也保证不了出的是对的qwq 可能会有错 原谅下qwq ）

**Reply by Dave-12138** [reply]
好像我的解释有重复定义的问题
>如果对方拿到18，对方猜自己应该是11、9
>11怪叫，9有2\*7不怪叫

这里发出b动静的是A。9不怪叫是因为如果A拿到9，发现对方可能拿到14=2\*7是送分题。
我认为A发现对方**不可能拿到送分题**时，A发出b动静。
B觉得A认为B不可能拿到送分题时发出b动静，因此用自己手中的数推测A可能拿到的数，并分析这些数是否满足条件，推得A拿到了11。
A看见B发出b动静，知道他通过自己的怪叫推出了自己拿到了11，但B拿到18、24、28都能通过A怪叫判断出A拿到11，A通过B的b动静只能拿到“B拿到的不是30”一个信息。

**Reply by qwq1949** [reply]
哦哦 是的qwq
那..是不是把A最后一句删了就行了qwq

**Reply by Dave-12138** [reply]
应该可以，话说好像A有几种情况可以在最后得到答案（并嘻嘻“我也知道了捏”），也就是满足没改的题目。
a|b|A|B
-- | -- | -- | --
4 | 13 | 17 | 52
4 | 61 | 65 | 244
4 | 181 | 185 | 724
16 | 73 | 89 | 1168
16 | 111 | 127 | 1776
16 | 163 | 179 | 2608
32 | 131 | 163 | 4192
64 | 73 | 137 | 4672

**Reply by qwq1949** [reply]
嗯嗯 改前的在查数列网站在1979 年 12 月的找到了 qwq

**Reply by Dave-12138** [reply]
给未来哪天再想起来这个的自己留个档
```javascript
/**
 * @link https://github.com/share121/inter-knot/discussions/809 
 */

const UPLIMIT = 114;
/**@type {Map<number,[number,number][]>} */
const rootsList = new Map();
/** 给定 a*b 获取可能的 a b */
function getRootSets(numB) {
    if (rootsList.has(numB)) {
        return rootsList.get(numB);
    }
    rootsList.set(numB, []);
    for (let a = 2; a < Math.sqrt(numB); a++) {
        if (numB % a !== 0) {
            continue;
        }
        const b = numB / a;
        rootsList.get(numB).push([a, b]);
    }
    return rootsList.get(numB);
}
/**@type {Map<number,[number,number][]>} */
const sumreasonList = new Map();
/** 给定 a+b 获取可能的 a b */
function getSumreasonSets(numA) {

    if (sumreasonList.has(numA)) {
        return sumreasonList.get(numA);
    }
    sumreasonList.set(numA, []);
    for (let a = 2; a <= numA / 2; a++) {
        const b = numA - a;
        sumreasonList.get(numA).push([a, b]);
    }
    return sumreasonList.get(numA);
}
// 缓存, 时间优化
/**
 * @template P,Q
 * @param {(v:P)=>Q} getter 
 * @param {Map<P,Q>} cacheMap 
 * @returns {(v:P)=>Q}
 */
function useCache(getter, cacheMap) {
    return function (v) {
        if (cacheMap.has(v)) {
            return cacheMap.get(v);
        }
        const rst = getter(v);
        cacheMap.set(v, rst);
        return rst;
    }
}
/**@type Map<number,boolean> */
const passed1 = new Map();
/**@type Map<number,boolean> */
const passed2 = new Map();
/**@type Map<number,boolean> */
const passed3 = new Map();
/*  对任意满足 a+b=numA 的组合 a,b , 都有 a*b=a1*b1=a2*b2 且a1!=b1!=a2!=b2 */
function check1(numA) {
    return useCache((numA) => getSumreasonSets(numA).every(([a, b]) => getRootSets(a * b).length > 1), passed1)(numA);
}
// 对满足 a*b=numB 的组合 a,b , 有唯一一组使得 a+b 满足check1
function check2(numB) {
    // 可以提前截断以优化
    return useCache(numB => getRootSets(numB).filter(([a, b]) => check1(a + b)).length == 1, passed2)(numB);
}
// 对满足 a+b=numA 的组合 a,b , 有唯一一组使得 a*b 满足check2
function check3(numA) {
    return useCache(numA => getSumreasonSets(numA).filter(([a, b]) => check2(a * b)).length == 1, passed3)(numA);
}
for (let a = 2; a < UPLIMIT; a++) {
    for (let b = a; b <= UPLIMIT; b++) {
        const A = a + b;
        const B = a * b;
        if (check1(A) && check2(B) && check3(A)) {
            console.log(`${a} ${b}`);
        }
    }
}
```

**Comment by xiaochenyu24** [comment]
我不道啊

**Comment by Asaki-333** [comment]
（⊙ｏ⊙）

**Comment by qwq1949** [comment]
改前的C++代码
```cpp
#include <bits/stdc++.h>
#define N 33333
#define M 222
using namespace std;

vector<int> p_sum; bool prime[N];
bool check1 ( int a )
{
	int pwq = 0;
	for ( int i = 2; i < sqrt(a); i ++ )
	if (!(a%i)) pwq ++;

	if ( pwq >= 2 ) return true;
	else return false;
}
int check2 ( int a )
{
	for ( int i = 2; ; i ++ )
	if (!check1(i*(a-i))) return false;
	else if ( i * 2 > a ) return true;
}
int main ()
{
	memset(prime, true, sizeof prime);
	prime[0] = prime[1] = false;
	
	for ( int i = 2; i * i < N; i ++ )
	if (prime[i]) for ( int j = i * i; j < N; j += i )
	prime[j] = false;
	
	for ( int i = 3; i < N; i += 2 )
	if (!prime[i]) p_sum.push_back(i+2);

	for ( int i = 2; i <= M; i ++ )
	for ( int j = i + 1; j <= M; j ++ )
	{
		if (!check2(i+j)) continue;
		int k = i * j, cnt = 0;

		for ( int o = 2; o <= sqrt(k); o ++ )
		if (!(k%o)&&check2(k/o+o)) cnt ++;
 		
		if ( cnt == 1 )
		{
			int he1 = i + j, f_cnt = 0;
			for ( int i1 = 2; i1 * 2 < he1; i1 ++ )
			{
				int ji1 = i1 * ( he1 - i1 ), p_cnt = 0;
				for ( int j1 = 2; j1 < sqrt(ji1); j1 ++ ) 
				if ( ji1 % j1 == 0 )
				{
					int he2 = ji1 / j1 + j1;
					for ( int kk = 0; kk < p_sum.size(); kk ++ )
					if ( he2 == p_sum[kk] ) p_cnt ++;
				} if ( p_cnt < 2 ) f_cnt ++; 
			} if ( f_cnt < 2 ) cout << i << " " << j << '\n';
		}
	}
	return 0;
}
```
改后的
```cpp
#include <bits/stdc++.h>
#define M 222
using namespace std;

bool check1 ( int a )
{
	int pwq = 0;
	for ( int i = 2; i < sqrt(a); i ++ )
	if (!(a%i)) pwq ++;

	if ( pwq >= 2 ) return true;
	else return false;
}
int check2 ( int a )
{
	for ( int i = 2; ; i ++ )
	if (!check1(i*(a-i))) return false;
	else if ( i * 2 > a ) return true;
}
int main ()
{
	for ( int i = 2; i <= M; i ++ )
	for ( int j = i + 1; j <= M; j ++ )
	{
		if (!check2(i+j)) continue;
		int k = i * j, cnt = 0;

		for ( int o = 2; o <= sqrt(k); o ++ )
		if (!(k%o)&&check2(k/o+o)) cnt ++;
 		
		if ( cnt == 1 )
		cout << i << " " << j << '\n';
	}
	return 0;
}

**Comment by YeMengXue** [comment]
// 回复“你是谁？”
console.log("你是谁？");
// 或者
function reply() {
&nbsp; &nbsp; return "你是谁？";
}


夜梦雪
***@***.***



        



         原始邮件
         
       
发件人：Dave_12138 ***@***.***&gt;
发件时间：2026年9月8日 20:23
收件人：share121/inter-knot ***@***.***&gt;
抄送：YeMengXue ***@***.***&gt;, Comment ***@***.***&gt;
主题：Re: [share121/inter-knot] 检测一下各位绳匠的编程能力 (Discussion #809)




给未来哪天再想起来这个的自己留个档
/**  * @link https://github.com/share121/inter-knot/discussions/809   */ const UPLIMIT = 114; ***@***.*** {Map<number,[number,number][]&gt;} */ const rootsList = new Map(); /** 给定 a*b 获取可能的 a b */ function getRootSets(numB) {  &nbsp; &nbsp;if (rootsList.has(numB)) {  &nbsp; &nbsp; &nbsp; &nbsp;return rootsList.get(numB);  &nbsp; &nbsp;}  &nbsp; &nbsp;rootsList.set(numB, []);  &nbsp; &nbsp;for (let a = 2; a < Math.sqrt(numB); a++) {  &nbsp; &nbsp; &nbsp; &nbsp;if (numB % a !== 0) {  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;continue;  &nbsp; &nbsp; &nbsp; &nbsp;}  &nbsp; &nbsp; &nbsp; &nbsp;const b = numB / a;  &nbsp; &nbsp; &nbsp; &nbsp;rootsList.get(numB).push([a, b]);  &nbsp; &nbsp;}  &nbsp; &nbsp;return rootsList.get(numB); } ***@***.*** {Map<number,[number,number][]&gt;} */ const sumreasonList = new Map(); /** 给定 a+b 获取可能的 a b */ function getSumreasonSets(numA) {  &nbsp; &nbsp;if (sumreasonList.has(numA)) {  &nbsp; &nbsp; &nbsp; &nbsp;return sumreasonList.get(numA);  &nbsp; &nbsp;}  &nbsp; &nbsp;sumreasonList.set(numA, []);  &nbsp; &nbsp;for (let a = 2; a <= numA / 2; a++) {  &nbsp; &nbsp; &nbsp; &nbsp;const b = numA - a;  &nbsp; &nbsp; &nbsp; &nbsp;sumreasonList.get(numA).push([a, b]);  &nbsp; &nbsp;}  &nbsp; &nbsp;return sumreasonList.get(numA); } // 缓存, 时间优化 ***@***.*** Map<number,boolean&gt; */ const passed1 = new Map(); ***@***.*** Map<number,boolean&gt; */ const passed2 = new Map(); ***@***.*** Map<number,boolean&gt; */ const passed3 = new Map(); /* &nbsp;对任意满足 a+b=numA 的组合 a,b , 都有 a*b=a1*b1=a2*b2 且a1!=b1!=a2!=b2 */ function check1(numA) {  &nbsp; &nbsp;if (passed1.has(numA)) {  &nbsp; &nbsp; &nbsp; &nbsp;return passed1.get(numA);  &nbsp; &nbsp;}  &nbsp; &nbsp;const rst = getSumreasonSets(numA).every(([a, b]) =&gt; getRootSets(a * b).length &gt; 1);  &nbsp; &nbsp;passed1.set(numA, rst);  &nbsp; &nbsp;return rst; } // 对满足 a*b=numB 的组合 a,b , 有唯一一组使得 a+b 满足check1 function check2(numB) {  &nbsp; &nbsp;if (passed2.has(numB)) {  &nbsp; &nbsp; &nbsp; &nbsp;return passed2.get(numB);  &nbsp; &nbsp;}  &nbsp; &nbsp;const rst = getRootSets(numB).filter(([a, b]) =&gt; check1(a + b)).length == 1; // 可以提前截断以优化  &nbsp; &nbsp;passed2.set(numB, rst);  &nbsp; &nbsp;return rst; } // 对满足 a+b=numA 的组合 a,b , 有唯一一组使得 a*b 满足check2 function check3(numA) {  &nbsp; &nbsp;if (passed3.has(numA)) {  &nbsp; &nbsp; &nbsp; &nbsp;return passed3.get(numA);  &nbsp; &nbsp;}  &nbsp; &nbsp;const rst = getSumreasonSets(numA).filter(([a, b]) =&gt; check2(a * b)).length == 1;  &nbsp; &nbsp;passed3.set(numA, rst);  &nbsp; &nbsp;return rst; } for (let a = 2; a < UPLIMIT; a++) {  &nbsp; &nbsp;for (let b = a; b <= UPLIMIT; b++) {  &nbsp; &nbsp; &nbsp; &nbsp;const A = a + b;  &nbsp; &nbsp; &nbsp; &nbsp;const B = a * b;  &nbsp; &nbsp; &nbsp; &nbsp;if (check1(A) &amp;&amp; check2(B) /* &amp;&amp; check3(A) */) {  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;console.log(`${a} ${b}`);  &nbsp; &nbsp; &nbsp; &nbsp;}  &nbsp; &nbsp;} }

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS&nbsp;and Android. Download it today!

**Reply by Dave-12138** [reply]
“You are receiving this because you commented.” 因为你在这个 GitHub 讨论里评论了，所以当别人在这里评论的时候会通知你。不在意就不用管，或者点那个`unsubscribe`取消订阅，或者在github设置关掉邮件通知

**Comment by 123acnl** [comment]
为什么绳网发不了帖子了啊
